cli/xml-to-ts.ts

```diff
@@ -1,69 +1,118 @@
-    pascalCase(filename.substring(0, filename.lastIndexOf(".")))
-  );
-  const xml = fs.readFileSync(filename, "utf-8");
+function generateInterface(xmlFilename: string, tsFilename: string) {
+  const interfaceName = xmltools.generateInterfaceName(tsFilename);
+  const xml = fs.readFileSync(xmlFilename, "utf-8");
   return xmltools.createInterface(interfaceName, xml);
 }
 
-function openTsFile(filename: string, flags: string): number {
-  filename = filename.substring(0, filename.lastIndexOf(".")) + ".ts";
-  return fs.openSync(filename, flags);
+// map<directory, suffix>
+const directorySuffix: { [key: string]: string } = {
+  parts: "-part-config",
+  pages: "-page-config",
+  site: "-config"
+};
+
+// tries to avoid conflicting filenames
+function getTsFilename(filename: string): string {
+  const dirname = path.dirname(filename);
+  const basename = path.basename(filename, path.extname(filename));
+
+  const closestDir: string = path
+    .dirname(filename)
+    .split(path.sep)
+    .reverse()
+    .slice(0, 2)
+    .find(p => directorySuffix[p]);
+  const suffix = directorySuffix[closestDir] || "";
+
+  return `${dirname}/${basename}${suffix}.ts`;
+}
+
+function replaceFileExtension(
+  newExtension: string
+): (filename: string) => string {
+  return filename =>
+    filename.substring(0, filename.length - path.extname(filename).length) +
+    newExtension;
 }
 
 // XML-files in these directories will generate TypeScript interfaces when
 // using the --enonic-xml flag.
 const directories = [
-  "./src/main/resources/site/site.xml",
-  "./src/main/resources/site/content-types",
-  "./src/main/resources/site/parts",
-  "./src/main/resources/site/pages"
+  "src/main/resources/site/site.xml",
+  "src/main/resources/site/content-types",
+  "src/main/resources/site/parts",
+  "src/main/resources/site/pages"
 ];
 
-function getEnonicXmlFiles() {
-  const files = [];
-  for (let dir of directories) {
-    dir = path.resolve(dir);
-    if (!fs.existsSync(dir)) {
-      continue;
-    }
+function getEnonicXmlFiles(projectRootDir: string): Array<string> {
+  return directories
+    .map(dir => path.join(projectRootDir, dir))
+    .map(dir => path.resolve(dir))
+    .filter(dir => fs.existsSync(dir))
+    .reduce((result: Array<string>, dir: string) => {
+      const stat = fs.statSync(dir);
+      return result
+        .concat(stat.isFile() ? [dir] : [])
+        .concat(stat.isDirectory() ? listXmlFiles(dir) : []);
+    }, []);
+}
 
-    const stat = fs.statSync(dir);
-    if (stat.isFile()) {
-      files.push(dir);
-    } else if (stat.isDirectory()) {
-      files.push(...fs.readdirSync(dir).map(filename => dir + filename));
-    }
-  }
-  return files;
+function listXmlFiles(dir: string): Array<string> {
+  return listFiles(dir).filter(f => path.extname(f) === ".xml");
+}
+
+function listFiles(dir: string): Array<string> {
+  return fs
+    .readdirSync(dir)
+    .map(file => path.join(dir, file))
+    .reduce((result, dir) => {
```

Comment:
I'm getting errors on line 76 if `result` doesn't have a type.
```suggestion
    .reduce((result: Array<string>, dir) => {
```